File: libs/server-nest/domain/src/domain/diagram/diagram.ts
import { Entity, HasMany, Ref } from '../core';
import { DiagramEdge, DiagramEdges } from './edge';
import { DiagramNode, DiagramNodes } from './node';
import { Viewport } from './types';

export interface DiagramDescription {
  workspace: Ref<string>;
  title: string;
  viewport: Viewport;
  createdAt: string;
  updatedAt: string;
}

export class Diagram implements Entity<string, DiagramDescription> {
  constructor(
    private readonly id: string,
    private readonly desc: DiagramDescription,
    private readonly diagramNodes: DiagramNodes,
    private readonly diagramEdges: DiagramEdges,
  ) {}

  identity(): string {
    return this.id;
  }

  description(): DiagramDescription {
    return this.desc;
  }

  nodes(): HasMany<DiagramNode> {
    return this.diagramNodes;
  }

  edges(): HasMany<DiagramEdge> {
    return this.diagramEdges;
  }
}

File: libs/server-java/domain/src/main/java/reengineering/ddd/evidence/domain/model/Diagram.java
package reengineering.ddd.evidence.domain.model;

import io.github.jayclock.smartdomain.core.Entity;
import io.github.jayclock.smartdomain.core.HasMany;
import reengineering.ddd.evidence.domain.description.DiagramDescription;

public final class Diagram implements Entity<String, DiagramDescription> {
  private final String identity;
  private final DiagramDescription description;
  private final HasMany<String, DiagramNode> nodes;
  private final HasMany<String, DiagramEdge> edges;

  public Diagram(
      String identity,
      DiagramDescription description,
      HasMany<String, DiagramNode> nodes,
      HasMany<String, DiagramEdge> edges) {
    this.identity = identity;
    this.description = description;
    this.nodes = nodes;
    this.edges = edges;
  }

  @Override
  public String getIdentity() {
    return identity;
  }

  @Override
  public DiagramDescription getDescription() {
    return description;
  }

  public HasMany<String, DiagramNode> nodes() {
    return nodes;
  }

  public HasMany<String, DiagramEdge> edges() {
    return edges;
  }

  public record Position(double x, double y) {}
}
